### PR TITLE
feat: support indent in region layout

### LINKED ISSUE #1876: [Bug] 设置了extendMax, 参考系依旧被裁剪
Image

Image

```
const spec = {
    "type": "line",
    "xField": [
        "220530101226439"
    ],
    "yField": [
        "10002"
    ],
    "direction": "vertical",
    "sortDataByAxis": true,
    "seriesField": "20001",
    "padding": 0,
    "labelLayout": "region",
    "data": [
        {
            "id": "data",
            "values": [
                {
                    "10001": "XXXX",
                    "10002": "-0.6073673361017367",
                    "10003": "220530104654532",
                    "20001": "XXXX",
                    "220530101226439": "2023-11-30",
                    "220530104654532": "-0.6073673361017367"
                },
                {
                    "10001": "XXXX",
                    "10002": "-0.6452117312689131",
                    "10003": "220530104654532",
                    "20001": "XXXX",
                    "220530101226439": "2023-12-01",
                    "220530104654532": "-0.6452117312689131"
                },
                {
                    "10001": "XXXX",
                    "10002": "-0.5112678570530983",
                    "10003": "220530104654532",
                    "20001": "XXXX",
                    "220530101226439": "2023-12-02",
                    "220530104654532": "-0.5112678570530983"
                },
                {
                    "10001": "XXXX",
                    "10002": "-0.5419587785754737",
                    "10003": "220530104654532",
                    "20001": "XXXX",
                    "220530101226439": "2023-12-03",
                    "220530104654532": "-0.5419587785754737"
                },
                {
                    "10001": "XXXX",
                    "10002": "-0.560423026652941",
                    "10003": "220530104654532",
                    "20001": "XXXX",
                    "220530101226439": "2023-12-04",
                    "220530104654532": "-0.560423026652941"
                },
                {
                    "10001": "XXXX",
                    "10002": "-0.5555956275132696",
                    "10003": "220530104654532",
                    "20001": "XXXX",
                    "220530101226439": "2023-12-05",
                    "220530104654532": "-0.5555956275132696"
                },
                {
                    "10001": "XXXX",
                    "10002": "-0.5686394576037566",
                    "10003": "220530104654532",
                    "20001": "XXXX",
                    "220530101226439": "2023-12-06",
                    "220530104654532": "-0.5686394576037566"
                },
                {
                    "10001": "XXXX",
                    "10002": "-0.5286660990110997",
                    "10003": "220530104654532",
                    "20001": "XXXX",
                    "220530101226439": "2023-12-07",
                    "220530104654532": "-0.5286660990110997"
                },
                {
                    "10001": "XXXX",
                    "10002": "-0.590044416648712",
                    "10003": "220530104654532",
                    "20001": "XXXX",
                    "220530101226439": "2023-12-08",
                    "220530104654532": "-0.590044416648712"
                },
                {
                    "10001": "XXXX",
                    "10002": "-0.5526256977959144",
                    "10003": "220530104654532",
                    "20001": "XXXX",
                    "220530101226439": "2023-12-09",
                    "220530104654532": "-0.5526256977959144"
                },
                {
                    "10001": "XXXX",
                    "10002": "-0.4990786147139401",
                    "10003": "220530104654532",
                    "20001": "XXXX",
                    "220530101226439": "2023-12-10",
                    "220530104654532": "-0.4990786147139401"
                },
                {
                    "10001": "XXXX",
                    "10002": "-0.5519450722616297",
                    "10003": "220530104654532",
                    "20001": "XXXX",
                    "220530101226439": "2023-12-11",
                    "220530104654532": "-0.5519450722616297"
                },
                {
                    "10001": "XXXX",
                    "10002": "-0.5324075796123078",
                    "10003": "220530104654532",
                    "20001": "XXXX",
                    "220530101226439": "2023-12-12",
                    "220530104654532": "-0.5324075796123078"
                },
                {
                    "10001": "XXXX",
                    "10002": "-0.5155552233504506",
                    "10003": "220530104654532",
                    "20001": "XXXX",
                    "220530101226439": "2023-12-13",
                    "220530104654532": "-0.5155552233504506"
                },
                {
                    "10001": "XXXX",
                    "10002": "-0.588308476818609",
                    "10003": "220530104654532",
                    "20001": "XXXX",
                    "220530101226439": "2023-12-14",
                    "220530104654532": "-0.588308476818609"
                },
                {
                    "10001": "XXXX",
                    "10002": "-0.5882418978998751",
                    "10003": "220530104654532",
                    "20001": "XXXX",
                    "220530101226439": "2023-12-15",
                    "220530104654532": "-0.5882418978998751"
                },
                {
                    "10001": "XXXX",
                    "10002": "-0.5624073967001346",
                    "10003": "220530104654532",
                    "20001": "XXXX",
                    "220530101226439": "2023-12-16",
                    "220530104654532": "-0.5624073967001346"
                },
                {
                    "10001": "XXXX",
                    "10002": "-0.5320075802074745",
                    "10003": "220530104654532",
                    "20001": "XXXX",
                    "220530101226439": "2023-12-17",
                    "220530104654532": "-0.5320075802074745"
                },
                {
                    "10001": "XXXX",
                    "10002": "-0.5399965619657939",
                    "10003": "220530104654532",
                    "20001": "XXXX",
                    "220530101226439": "2023-12-18",
                    "220530104654532": "-0.5399965619657939"
                },
                {
                    "10001": "XXXX",
                    "10002": "-0.45722063570153526",
                    "10003": "220530104654532",
                    "20001": "XXXX",
                    "220530101226439": "2023-12-19",
                    "220530104654532": "-0.45722063570153526"
                },
                {
                    "10001": "XXXX",
                    "10002": "-0.44863925953359424",
                    "10003": "220530104654532",
                    "20001": "XXXX",
                    "220530101226439": "2023-12-20",
                    "220530104654532": "-0.44863925953359424"
                },
                {
                    "10001": "XXXX",
                    "10002": "-0.5681363573737701",
                    "10003": "220530104654532",
                    "20001": "XXXX",
                    "220530101226439": "2023-12-21",
                    "220530104654532": "-0.5681363573737701"
                },
                {
                    "10001": "XXXX",
                    "10002": "-0.5102980490450119",
                    "10003": "220530104654532",
                    "20001": "XXXX",
                    "220530101226439": "2023-12-22",
                    "220530104654532": "-0.5102980490450119"
                },
                {
                    "10001": "XXXX",
                    "10002": "-0.5074579755548064",
                    "10003": "220530104654532",
                    "20001": "XXXX",
                    "220530101226439": "2023-12-23",
                    "220530104654532": "-0.5074579755548064"
                },
                {
                    "10001": "XXXX",
                    "10002": "-0.4480726680112404",
                    "10003": "220530104654532",
                    "20001": "XXXX",
                    "220530101226439": "2023-12-24",
                    "220530104654532": "-0.4480726680112404"
                },
                {
                    "10001": "XXXX",
                    "10002": "-0.4252908123072075",
                    "10003": "220530104654532",
                    "20001": "XXXX",
                    "220530101226439": "2023-12-25",
                    "220530104654532": "-0.4252908123072075"
                },
                {
                    "10001": "XXXX",
                    "10002": "-0.4179822827499412",
                    "10003": "220530104654532",
                    "20001": "XXXX",
                    "220530101226439": "2023-12-26",
                    "220530104654532": "-0.4179822827499412"
                },
                {
                    "10001": "XXXX",
                    "10002": "-0.463012439312728",
                    "10003": "220530104654532",
                    "20001": "XXXX",
                    "220530101226439": "2023-12-27",
                    "220530104654532": "-0.463012439312728"
                },
                {
                    "10001": "XXXX",
                    "10002": null,
                    "10003": "220530104654532",
                    "20001": "XXXX",
                    "220530101226439": "2023-12-28",
                    "220530104654532": null
                }
            ],
            "fields": {
                "10001": {
                    "alias": "指标名称 "
                },
                "10002": {
                    "alias": "指标值 "
                },
                "20001": {
                    "alias": "图例项 ",
                    "domain": [
                        "XXXX"
                    ],
                    "sortIndex": 0,
                    "lockStatisticsByDomain": true
                },
                "220530101226439": {
                    "alias": "日期",
                    "domain": [
                        "2023-11-30",
                        "2023-12-01",
                        "2023-12-02",
                        "2023-12-03",
                        "2023-12-04",
                        "2023-12-05",
                        "2023-12-06",
                        "2023-12-07",
                        "2023-12-08",
                        "2023-12-09",
                        "2023-12-10",
                        "2023-12-11",
                        "2023-12-12",
                        "2023-12-13",
                        "2023-12-14",
                        "2023-12-15",
                        "2023-12-16",
                        "2023-12-17",
                        "2023-12-18",
                        "2023-12-19",
                        "2023-12-20",
                        "2023-12-21",
                        "2023-12-22",
                        "2023-12-23",
                        "2023-12-24",
                        "2023-12-25",
                        "2023-12-26",
                        "2023-12-27",
                        "2023-12-28"
                    ],
                    "sortIndex": 0,
                    "lockStatisticsByDomain": true
                },
                "220530104654532": {
                    "alias": "XXXX"
                },
                "220530104654533": {
                    "alias": "XXXX的环比增长值"
                },
                "220530104654534": {
                    "alias": "XXXX的环比原始值"
                }
            }
        }
    ],
    "stackInverse": true,
    "axes": [
        {
            "type": "band",
            "tick": {
                "visible": false
            },
            "grid": {
                "visible": false,
                "style": {
                    "zIndex": 150,
                    "stroke": "#DADCDD",
                    "lineWidth": 1,
                    "lineDash": [
                        4,
                        2
                    ]
                }
            },
            "orient": "bottom",
            "visible": true,
            "domainLine": {
                "visible": true,
                "style": {
                    "lineWidth": 1,
                    "stroke": "#989999"
                }
            },
            "title": {
                "visible": false,
                "space": 5,
                "text": "日期",
                "style": {
                    "fontSize": 8,
                    "fill": "#363839",
                    "fontWeight": "normal"
                }
            },
            "sampling": false,
            "zIndex": 200,
            "label": {
                "visible": true,
                "space": 4,
                "style": {
                    "fontSize": 8,
                    "fill": "#6F6F6F",
                    "angle": 0,
                    "fontWeight": "normal",
                    "direction": "horizontal",
                    "maxLineWidth": 174
                },
                "autoHide": true,
                "autoHideMethod": "greedy",
                "flush": true
            },
            "hover": true,
            "background": {
                "visible": true,
                "state": {
                    "hover": {
                        "fillOpacity": 0.08,
                        "fill": "#141414"
                    },
                    "hover_reverse": {
                        "fillOpacity": 0.08,
                        "fill": "#141414"
                    }
                }
            },
            "paddingInner": 0.36249999999999993,
            "paddingOuter": 0.175
        },
        {
            "type": "linear",
            "tick": {
                "visible": false,
                "tickMode": "d3",
                "style": {
                    "stroke": "rgba(255, 255, 255, 0)"
                }
            },
            "niceType": "accurateFirst",
            "zIndex": 200,
            "grid": {
                "visible": true,
                "style": {
                    "zIndex": 150,
                    "stroke": "#DADCDD",
                    "lineWidth": 1,
                    "lineDash": [
                        4,
                        2
                    ]
                }
            },
            "orient": "left",
            "visible": true,
            "domainLine": {
                "visible": true,
                "style": {
                    "lineWidth": 1,
                    "stroke": "rgba(255, 255, 255, 0)"
                }
            },
            "title": {
                "visible": false,
                "text": "XXXX",
                "space": 8,
                "style": {
                    "fontSize": 8,
                    "fill": "#363839",
                    "fontWeight": "normal"
                }
            },
            "sampling": false,
            "label": {
                "visible": true,
                "space": 6,
                "flush": true,
                "padding": 0,
                "style": {
                    "fontSize": 8,
                    "maxLineWidth": 174,
                    "fill": "#6F6F6F",
                    "angle": 0,
                    "fontWeight": "normal",
                    "dy": -1,
                    "direction": "horizontal"
                },
                "autoHide": true,
                "autoHideMethod": "greedy"
            },
            "background": {
                "visible": true,
                "state": {
                    "hover": {
                        "fillOpacity": 0.08,
                        "fill": "#141414"
                    },
                    "hover_reverse": {
                        "fillOpacity": 0.08,
                        "fill": "#141414"
                    }
                }
            },
            "min": null,
            "max": null,
            "zero": true,
            "nice": false,
            "expand": {
                "max": 0.017
            }
        }
    ],
    "color": {
        "field": "20001",
        "type": "ordinal",
        "range": [
            "#2E62F1"
        ],
        "specified": {},
        "domain": [
            "XXXX"
        ]
    },
    "legends": [
        {
            "type": "discrete",
            "visible": false
        }
    ],
    "label": {
        "visible": false,
        "offset": 3,
        "overlap": {
            "hideOnHit": true,
            "avoidBaseMark": false,
            "strategy": [
                {
                    "type": "position",
                    "position": [
                        "top",
                        "bottom"
                    ]
                }
            ],
            "clampForce": true
        },
        "style": {
            "fontSize": 12,
            "fontWeight": "normal",
            "zIndex": 400,
            "lineHeight": "100%",
            "fill": "#363839",
            "stroke": "rgba(255, 255, 255, 0.8)",
            "lineWidth": 1.5,
            "strokeOpacity": 1
        },
        "position": "top",
        "smartInvert": false
    },
    "tooltip": {
        "handler": {}
    },
    "point": {
        "style": {
            "shape": {
                "type": "ordinal",
                "field": "20001",
                "range": [
                    "circle"
                ],
                "domain": [
                    "XXXX"
                ]
            },
            "size": {
                "type": "ordinal",
                "field": "20001",
                "range": [
                    3.5449077018110318
                ],
                "domain": [
                    "XXXX"
                ]
            },
            "fill": "white",
            "stroke": {
                "field": "20001",
                "type": "ordinal",
                "range": [
                    "#2E62F1"
                ],
                "specified": {},
                "domain": [
                    "XXXX"
                ]
            },
            "strokeOpacity": {
                "type": "ordinal",
                "field": "20001",
                "range": [
                    0
                ],
                "domain": [
                    "XXXX"
                ]
            },
            "fillOpacity": {
                "type": "ordinal",
                "field": "20001",
                "range": [
                    0
                ],
                "domain": [
                    "XXXX"
                ]
            }
        },
        "state": {
            "hover": {
                "lineWidth": 2,
                "fillOpacity": 1,
                "strokeOpacity": 1,
                "scaleX": 1.5,
                "scaleY": 1.5
            }
        }
    },
    "line": {
        "style": {
            "curveType": {
                "type": "ordinal",
                "field": "20001",
                "range": [
                    "monotoneX"
                ],
                "domain": [
                    "XXXX"
                ]
            },
            "lineWidth": {
                "type": "ordinal",
                "field": "20001",
                "range": [
                    1
                ],
                "domain": [
                    "XXXX"
                ]
            },
            "lineDash": {
                "type": "ordinal",
                "field": "20001",
                "range": [
                    [
                        0,
                        0
                    ]
                ],
                "domain": [
                    "XXXX"
                ]
            }
        }
    },
    "seriesMark": "line",
    "markOverlap": false,
    "region": [
        {
            "clip": true
        }
    ],
    "background": "rgba(255, 255, 255, 0)",
    "area": {
        "style": {
            "curveType": {
                "type": "ordinal",
                "field": "20001",
                "range": [
                    "monotoneX"
                ],
                "domain": [
                    "XXXX"
                ]
            }
        }
    },
    "invalidType": "break",
    "markLine": [
        {
            "y": 0.05,
            "startSymbol": {
                "visible": true,
                "symbolType": "triangleDown",
                "style": {
                    "size": 10,
                    "dx": 5,
                    "fill": "#f3a016"
                }
            },
            "endSymbol": {
                "visible": false
            },
            "autoRange": true,
            "label": {
                "visible": true,
                "style": {
                    "dx": -4,
                    "dy": 0,
                    "fontSize": 12,
                    "fontWeight": "normal",
                    "fill": "#fff",
                    "cursor": "pointer"
                },
                "confine": true,
                "position": "insideStartTop",
                "labelBackground": {
                    "visible": true,
                    "padding": {
                        "left": 5,
                        "right": 5,
                        "top": 2,
                        "bottom": 2
                    },
                    "style": {
                        "fill": "#2F3B52",
                        "fillOpacity": 0.9,
                        "dx": -4,
                        "dy": 0
                    }
                }
            },
            "line": {
                "style": {
                    "stroke": "#f3a016",
                    "lineWidth": 2,
                    "lineDash": [
                        3,
                        3
                    ],
                    "cursor": "pointer"
                }
            },
            "relativeSeriesId": "mainSeries",
            "id": "08d81e1e-0b30-4146-aaa6-40ab31163f4a",
            "interactive": true
        },
        {
            "y": -0.05,
            "startSymbol": {
                "visible": true,
                "symbolType": "triangleDown",
                "style": {
                    "size": 10,
                    "dx": 5,
                    "fill": "#f3a016"
                }
            },
            "endSymbol": {
                "visible": false
            },
            "autoRange": true,
            "label": {
                "visible": true,
                "style": {
                    "dx": -4,
                    "dy": 0,
                    "fontSize": 12,
                    "fontWeight": "normal",
                    "fill": "#fff",
                    "cursor": "pointer"
                },
                "confine": true,
                "position": "insideStartTop",
                "labelBackground": {
                    "visible": true,
                    "padding": {
                        "left": 5,
                        "right": 5,
                        "top": 2,
                        "bottom": 2
                    },
                    "style": {
                        "fill": "#2F3B52",
                        "fillOpacity": 0.9,
                        "dx": -4,
                        "dy": 0
                    }
                }
            },
            "line": {
                "style": {
                    "stroke": "#f3a016",
                    "lineWidth": 2,
                    "lineDash": [
                        3,
                        3
                    ],
                    "cursor": "pointer"
                }
            },
            "relativeSeriesId": "mainSeries",
            "id": "77f8c73e-cdc1-48c3-9fec-b0995131da50",
            "interactive": true
        }
    ],
    "animation": false,
    "brush": {
        "inBrush": {
            "fillOpacity": 1,
            "stroke": "#58595B",
            "lineWidth": 1,
            "colorAlpha": 1
        },
        "outOfBrush": {
            "colorAlpha": 0.2,
            "fillOpacity": 0.3,
            "strokeWidth": 0.3
        }
    },
    "crosshair": {
        "xField": {
            "visible": true
        },
        "gridZIndex": 100
    },
    "hash": "6d0d0d396f521f9230f239e163a94db3"
}

const vchart = new VChart(spec, { dom: CONTAINER_ID });
vchart.renderAsync();

// Just for the convenience of console debugging, DO NOT COPY!
window['vchart'] = vchart;
```

### PR DESCRIPTION
**Pull request by xuefei1313**
[[中文版模板 / Chinese template](https://github.com/VisActor/VChart/blob/main/.github/PULL_REQUEST_TEMPLATE/pr_cn.md?plain=1)]

### 🤔 This is a ...

- [X] New feature
- [ ] Bug fix
- [ ] TypeScript definition update
- [ ] Bundle size optimization
- [ ] Performance optimization
- [ ] Enhancement feature
- [ ] Refactoring
- [ ] Update dependency
- [ ] Code style optimization
- [ ] Test Case
- [ ] Branch merge
- [ ] Release
- [ ] Site / documentation update
- [ ] Demo update
- [ ] Workflow
- [ ] Other (about what?)

### 🔗 Related issue link

close #1876 #1234 

### 📝 Changelog

| Language   | Changelog |
| ---------- | --------- |
| 🇺🇸 English |           |
| 🇨🇳 Chinese |           |

### ☑️ Self-Check before Merge

⚠️ Please check all items below before requesting a reviewing. ⚠️

- [ ] Doc is updated/provided or not needed
- [ ] Demo is updated/provided or not needed
- [ ] TypeScript definition is updated/provided or not needed
- [ ] Changelog is provided or not needed

---

### 🚀 Summary

copilot:summary

### 🔍 Walkthrough

copilot:walkthrough

### CODE REVIEW

File: packages/vchart/src/model/interface.ts [hunk 1]
```diff
@@ -141,6 +141,11 @@ export interface ILayoutModel extends IModel {
   getLayoutRect: () => ILayoutRect;
   setLayoutRect: (rect: Partial<ILayoutRect>, levelMap?: Partial<ILayoutRect>) => void;
 
+  // 得到排除锁进后的布局起点
+  getLayoutPositionWidthIndent: () => IPoint;
```
Comment (xile611) [inline packages/vchart/src/model/interface.ts:145]: 这个函数命名是不是有问题？ `getLayoutPositionWithIndent` ?下面也是一样的

File: packages/vchart/src/layout/layout-item.ts [hunk 2]
```diff
@@ -285,6 +314,9 @@ export class LayoutItem implements ILayoutItem {
     }
 
     this.setRectInSpec(this._layoutRect);
+
+    this._layoutWithIndent.width = this._layoutRect.width - this._indent.left - this._indent.right;
```
Comment (xile611) [inline packages/vchart/src/layout/layout-item.ts:318]: 这里有可能会有负数的情况吧。。。

Comment (xuefei1313) [reply packages/vchart/src/layout/layout-item.ts:318]: 确实，等我兼容一下，至少有1px？0的话我担心会有计算上的问题

File: packages/vchart/src/model/interface.ts (hunk 1 above)
Comment (xuefei1313) [reply packages/vchart/src/model/interface.ts:145]: 。。。 尴尬，我直接改成 xxxExcludIndent 吧

**Review by xile611 — APPROVED** [review]

File: packages/vchart/src/component/marker/utils.ts [hunk 3]
```diff
@@ -264,17 +264,17 @@ export function computeClipRange(regions: IRegion[]) {
   let minY = Infinity;
   let maxY = -Infinity;
   regions.forEach((region: IRegion) => {
-    if (region.getLayoutStartPoint().x < minX) {
-      minX = region.getLayoutStartPoint().x;
+    if (region.getLayoutPositionExcludeIndent().x < minX) {
```
Comment (xile611) [inline packages/vchart/src/component/marker/utils.ts:267]: 这里会调用多次 `getLayoutPositionExcludeIndent()` 结果是不是应该保存一下

File: packages/vchart/src/component/marker/utils.ts [hunk 4]
```diff
@@ -222,8 +222,8 @@ export function coordinateLayout(
 export function positionLayout(positions: MarkerPositionPoint[], series: ISeries, regionRelative: boolean): IPoint[] {
   if (regionRelative) {
     const region = series.getRegion();
-    const { x: regionStartX, y: regionStartY } = region.getLayoutStartPoint();
-    const { width: regionWidth, height: regionHeight } = region.getLayoutRect();
+    const { x: regionStartX, y: regionStartY } = region.getLayoutPositionExcludeIndent();
```
Comment (xile611) [inline packages/vchart/src/component/marker/utils.ts:225]: 多次调用 `getLayoutPositionExcludeIndent()` 结果最好保存一下

**Comment by xile611** [comment]
@xuefei1313 文档和demo最好更新一下，目测这个功能应该不太会变化了吧

**Review by xile611 — APPROVED** [review]

**Pull request merged**